x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/DocumentPermissions.java

```diff
@@ -240,49 +227,44 @@ public static DocumentPermissions allowAll() {
      * @return instance of {@link DocumentPermissions}
      */
     public DocumentPermissions limitDocumentPermissions(DocumentPermissions limitedByDocumentPermissions) {
-        assert limitedByQueries == null && limitedByDocumentPermissions.limitedByQueries == null
-            : "nested scoping for document permissions is not permitted";
-        if (queries == null && limitedByDocumentPermissions.queries == null) {
+        if (hasDocumentLevelPermissions() && limitedByDocumentPermissions.hasDocumentLevelPermissions()) {
+            return new DocumentPermissions(
+                Stream.concat(getListOfQueries().stream(), limitedByDocumentPermissions.getListOfQueries().stream()).toList()
+            );
+        } else if (hasDocumentLevelPermissions()) {
+            return new DocumentPermissions(getListOfQueries());
+        } else if (limitedByDocumentPermissions.hasDocumentLevelPermissions()) {
+            return new DocumentPermissions(limitedByDocumentPermissions.getListOfQueries());
+        } else {
             return DocumentPermissions.allowAll();
         }
-        // TODO: should we apply the same logic here as FieldPermissions#limitFieldPermissions,
-        // i.e. treat limited-by as queries if original queries is null?
-        return new DocumentPermissions(queries, limitedByDocumentPermissions.queries);
     }
 
     @Override
     public String toString() {
-        return "DocumentPermissions [queries=" + queries + ", scopedByQueries=" + limitedByQueries + "]";
+        return "DocumentPermissions [listOfQueries=" + listOfQueries + "]";
     }
 
     @Override
     public void buildCacheKey(StreamOutput out, DlsQueryEvaluationContext context) throws IOException {
-        assert false == (queries == null && limitedByQueries == null) : "one of queries and limited-by queries must be non-null";
-        evaluateQueries(context);
-        if (evaluatedQueries != null) {
-            out.writeBoolean(true);
-            out.writeCollection(evaluatedQueries, StreamOutput::writeString);
-        } else {
-            out.writeBoolean(false);
-        }
-        if (evaluatedLimitedByQueries != null) {
-            out.writeBoolean(true);
-            out.writeCollection(evaluatedLimitedByQueries, StreamOutput::writeString);
-        } else {
-            out.writeBoolean(false);
+        if (false == hasDocumentLevelPermissions()) {
+            assert false;
```

Comment:
as mentioned in person , probably more of style thing, but I not a fan of both assertions and throwing. If you keep both, can you add a comment to the assertion.

Reply:
I changed it to just the assertion since the existing code was just an assertion. Thanks!